model_compression_toolkit/core/pytorch/back2framework/model_gradients.py

```diff
@@ -0,0 +1,256 @@
+        out_tensors_of_n = op_func(input_tensors, *op_call_args, **functional_kwargs)
+    else:
+        out_tensors_of_n = op_func(*input_tensors + op_call_args, **functional_kwargs)
+
+    return out_tensors_of_n
+
+
+def generate_outputs(
+        out_nodes: List[BaseNode],
+        node_to_output_tensors_dict: dict):
+    """
+    Args:
+        out_nodes: List of output nodes.
+        node_to_output_tensors_dict: A dictionary from a node to its output tensors.
+    Returns: List of output tensor/s for the model
+    """
+    output = []
+    for n in out_nodes:
+        out_tensors_of_n = node_to_output_tensors_dict.get(n)
+        if len(out_tensors_of_n) > 1:
+            output.append(out_tensors_of_n)
+        else:
+            output += out_tensors_of_n
+    return output
+
+
+class PytorchModelGradients(torch.nn.Module):
+    """
+    A Pytorch Module class for producing differentiable outputs from a given model's graph representation.
+    """
+    def __init__(self,
+                 graph_float: common.Graph,
+                 interest_points: List[BaseNode]):
+        """
+        Construct a PytorchModelGradients model.
+
+        Args:
+            graph_float: Model's Graph representation to evaluate the outputs according to.
+            interest_points: List of nodes in the graph which we want to produce outputs for.
+        """
+
+        super(PytorchModelGradients, self).__init__()
+        self.graph_float = graph_float
+        self.node_sort = list(topological_sort(graph_float))
+        self.interest_points = interest_points
+
+        for n in self.node_sort:
+            if not isinstance(n, FunctionalNode):
+                self.add_module(n.name, node_builder(n))
+
+    def forward(self,
+                *args: Any) -> Any:
+        """
+        Args:
+            args: argument input tensors to model, which is a mappings between an input node and its input tensor.
+
+        Returns:
+            A list of output tensors for each of the model's pre-defined interest points.
+        """
+
+        input_node_to_input_tensor = args[0]
+        node_to_output_tensors_dict = dict()
+        for n in self.node_sort:
+            input_tensors = build_input_tensors_list(n,
+                                                     self.graph_float,
+                                                     input_node_to_input_tensor,
+                                                     node_to_output_tensors_dict)
+
+            op_func = n.type if isinstance(n, FunctionalNode) else getattr(self, n.name)
+            out_tensors_of_n = run_operation(n,  # Run node operation and fetch outputs
+                                             input_tensors,
+                                             op_func=op_func)
+
+            if isinstance(out_tensors_of_n, list):
+                output_t = []
+                for t in out_tensors_of_n:
+                    t.retain_grad()
```

Comment:
1. Why we don't check here if the tensor requires_grad?
2. Shouldn't we use gradients only for interest points?